collector/lib/CollectorConnectionStats.h

```diff
@@ -0,0 +1,61 @@
+#ifndef _COLLECTOR_CONNECTION_STATS_
+#define _COLLECTOR_CONNECTION_STATS_
+
+#include <algorithm>
+
+#include "prometheus/gauge.h"
+#include "prometheus/registry.h"
+#include "prometheus/summary.h"
+
+namespace collector {
+template <typename T>
+class CollectorConnectionStats {
+ public:
+  CollectorConnectionStats(
+      prometheus::Registry* registry,
+      const std::string& name,
+      const std::string& help,
+      std::chrono::milliseconds max_age,
+      const std::vector<double>& quantiles,
+      double error) {
+    auto& family = prometheus::BuildSummary()
+                       .Name(name)
+                       .Help(help)
+                       .Register(*registry);
+    auto q = MakeQuantiles(quantiles, error);
+    inbound_private_summary_ = &family.Add({{"dir", "in"}, {"peer", "private"}}, q, max_age);
+    inbound_public_summary_ = &family.Add({{"dir", "in"}, {"peer", "public"}}, q, max_age);
+    outbound_private_summary_ = &family.Add({{"dir", "out"}, {"peer", "private"}}, q, max_age);
+    outbound_public_summary_ = &family.Add({{"dir", "out"}, {"peer", "public"}}, q, max_age);
+  }
+
+  void Observe(T inbound_private, T inbound_public, T outbound_private, T outbound_public) {
+    inbound_private_summary_->Observe(inbound_private);
+    inbound_public_summary_->Observe(inbound_public);
+    outbound_private_summary_->Observe(outbound_private);
+    outbound_public_summary_->Observe(outbound_public);
+  }
+
+ private:
+  prometheus::Summary* inbound_private_summary_;
+  prometheus::Summary* inbound_public_summary_;
+  prometheus::Summary* outbound_private_summary_;
+  prometheus::Summary* outbound_public_summary_;
```

Comment:
These have been turned from references to raw pointers. The way they are created makes it so the prometheus registry owns the actual objects and we only get a reference, however, storing references implicitly makes it so the copy and move constructors and operators need to be deleted, (because references are non-nullable and they can't be assigned to point into the ether).

Storing the raw pointers makes handling of the class easier and should be safe as long as we stop the network notifier before we destroy the prometheus exposer (which we do, because they are both owned by the CollectorService object and the exposer is stopped in its destructor).